connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -1515,6 +1520,79 @@ public void connectorOffsets(String connName, Callback<ConnectorOffsets> cb) {
+                        // We need to ensure that we perform the necessary checks again before proceeding to actually altering the connector offsets since
+                        // zombie fencing is done asynchronously and the conditions could have changed since the previous check
+                        addRequest(() -> {
+                            if (alterConnectorOffsetsChecks(connName, callback)) {
+                                worker.alterConnectorOffsets(connName, configState.connectorConfig(connName), offsets, callback);
+                            }
+                            return null;
+                        }, forwardErrorCallback(callback));
+                    }
+                });
+            } else {
+                worker.alterConnectorOffsets(connName, configState.connectorConfig(connName), offsets, callback);
+            }
+            return null;
+        }, forwardErrorCallback(callback));
+    }
+
+    /**
+     * This method performs a few checks for alter connector offsets request and completes the callback exceptionally
+     * if any check fails.
+     * @param connName the name of the connector whose offsets are to be altered
+     * @param callback callback to invoke upon completion
+     * @return true if all the checks passed, false otherwise
+     */
+    private boolean alterConnectorOffsetsChecks(String connName, Callback<Message> callback) {
+        if (checkRebalanceNeeded(callback)) {
+            return false;
+        }
+
+        if (!isLeader()) {
+            callback.onCompletion(new NotLeaderException("Only the leader can process alter offsets requests", leaderUrl()), null);
+            return false;
+        }
+
+        if (!refreshConfigSnapshot(workerSyncTimeoutMs)) {
+            throw new ConnectException("Failed to read to end of config topic before altering connector offsets");
+        }
+
+        if (!configState.contains(connName)) {
+            callback.onCompletion(new NotFoundException("Connector " + connName + " not found", null), null);
+            return false;
+        }
+
+        // If the target state for the connector is stopped, its task count is 0, and there is no rebalance pending (checked above),
+        // we can be sure that the tasks have at least been attempted to be stopped (or cancelled if they took too long to stop).
+        // Zombie tasks are handled by a round of zombie fencing for exactly once source connectors. Zombie sink tasks are handled
+        // naturally because requests to alter consumer group offsets will fail if there are still active members in the group.
+        if (configState.targetState(connName) != TargetState.STOPPED || configState.taskCount(connName) != 0) {
+            callback.onCompletion(new BadRequestException("The connector needs to be stopped before its offsets can be altered"), null);
```

Comment:
We should be really careful about the wording in this message (and the one in `StandaloneHerder`), since "stopping" a connector used to be synonymous with "deleting" one.

A quick sketch of a more-informative error message:
> "Connectors must be in the STOPPED state before their offsets can be altered. This can be done for the specified connector by issuing a PUT request to the /connectors/" + connName + "/stop endpoint.